File: product-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
    //tc: O(n) sc: O(n)
        int[] result=new int[nums.length];
        int[] left=new int[nums.length];
        int[] right=new int[nums.length];
        int prod=1;
        left[0]=1;
        for(int i=1;i<nums.length;i++){
            prod=prod*nums[i-1];
            left[i]=prod;
        }
        
        prod=1;
        right[right.length-1]=1;
        for(int j=nums.length-2;j>=0;j--){
            prod=prod*nums[j+1];
            right[j]=prod;
        }
        for(int i=0;i<nums.length;i++){
            result[i]=left[i]*right[i];
        }
        return result;
    }
}

File: productself.py
'''
Time Complexity: O(N)
Space Complexity: O(1)
'''
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        rp1 = 1
        rp2 = 1
        temp_arr = []
        for i in range(0, len(nums)):
            if(i==0):
                temp_arr.append(rp1)
            else:
                rp1 = rp1*nums[i-1]
                temp_arr.append(rp1)
        for i in range(len(temp_arr)-2, -1, -1):
            rp2 = rp2*nums[i+1]
            temp_arr[i] = temp_arr[i]*rp2
        return temp_arr
        